src/java/org/apache/cassandra/index/sai/disk/vector/OnDiskOrdinalsMap.java

```diff
@@ -227,21 +229,59 @@ public void close()
         }
     }
 
+    /**
+     * not thread safe
+     */
     private class FileReadingOrdinalsView implements OrdinalsView
     {
         RandomAccessReader reader = fh.createReader();
         private final long high = (segmentEnd - 8 - rowOrdinalOffset) / 8;
+        private int lastFoundRowId = -1;
+        private long lastFoundRowIdIndex = -1;
 
         /**
          * @return order if given row id is found; otherwise return -1
          */
         @Override
         public int getOrdinalForRowId(int rowId) throws IOException
         {
+            if (rowId <= lastFoundRowId)
+                logger.warn("repeated read for rowId {}, last found rowid: {}", rowId, lastFoundRowId);
+
+            long low = 0;
+            if (lastFoundRowId > -1 && lastFoundRowId <= rowId && lastFoundRowIdIndex < high)
+            {
+                low = lastFoundRowIdIndex;
+
+                if (low == rowId) //repeated read, not really expected
```

Comment:
I think we want to check `lastFoundRowId == rowId`. It is meaningful in cases where the `lastFoundRowId` skipped some number of rows. I agree this case is not commonly expected, but it should be cheap to check.